Disable SDS if is not supported

lucmaga opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/pmdk/4896)
<!-- Reviewable:end -->

lucmaga commented:
@pbalcer I stumble in the obj_sds test. It mock ndctrl calls so the SDS will not be disabled as it should be.
Any ideas?

pbalcer commented:
Correctly mocking ndctl will be difficult. I suggest just mocking the `pmem2_source_device_usc` and `pmem2_source_device_id` functions instead.

I'm not sure if hooking into incompat feature is the right choice for this - @lplewa?

lucmaga commented:
@pbalcer Sorry, I was not clean on my last comment.
The obj_sds test mocks the `pmem2_source_device_usc` and `pmem2_source_device_id` calls. This brokes the mechanism to test if SDS is supported as I use `pmem2_source_device_usc` for this.

pbalcer commented:
Right, I looked at the mocks now and I see what you mean.
Just return 0 on the first call to `pmem2_source_device_usc` mock, this will make `shutdown_state_is_supported` return true.

pbalcer commented on src/test/obj_sds/obj_sds.c:
```diff
@@ -91,6 +91,9 @@ FUNC_MOCK(pmem2_source_device_id, int, const struct pmem2_source *src,
 FUNC_MOCK_END
 FUNC_MOCK(pmem2_source_device_usc, int, const struct pmem2_source *src,
 		uint64_t *usc)
+	FUNC_MOCK_RUN(1){
```
this should probably be 0.

lucmaga replied:
Yeah, even so it is still failing.

lplewa commented on src/libpmem2/usc_ndctl.c:
```diff
@@ -50,7 +50,7 @@ pmem2_source_device_usc(const struct pmem2_source *src, uint64_t *usc)
 
 	if (region == NULL) {
-		ERR(
+		LOG(3,
```
if pmem2 returns error it should set ERR

lucmaga commented:
The pmempool_feature and libpmempool_feature tests that are SDS related are being disabled on the CI by this patchset.
Maybe this will need a workaround?

pbalcer commented:
I think that's fine, we run all PMDK tests on machines that have SDS anyway.

lucmaga commented:
Hi, I need some help with this windows tests, in special the `obj_sds/TEST2`. I don't
have access to a windows machine :(.

It's working correctly on Linux but broken on windows. I might be missing something
about this mocked functions that need to be done on windows.

I also notice that the vcxproj files force include many .c files on the build. Why is that so?
Isn't link the library enough?The including of `src/common/shutdown_state.c` was causing
a double definition of `shutdown_state_is_supported` and braking the build.

pbalcer commented on src/common/shutdown_state.c:
```diff
@@ -20,6 +20,29 @@
+/*
+ * shutdown_state_is_supported -- (internal) check if device supports SDS
+ *
+ * Returns 0 if suppots and 1 if not.
```
supports

pbalcer commented:
I tried to look at the code to try to think of a reason for why this might be failing on windows, but I haven't come up with anything sensible.
@lplewa can you please help?

lucmaga commented:
It seams that the neither `__wrap_pmem2_source_device_id` or `__wrap_pmem2_source_device_usc` is being called. I added
and output on the mocked functions but only the `__wrap_shutdown_state_is_supported` is printing out.

You can see the output
```
obj_sds/TEST2: SETUP (check\pmem\debug)
The fist run failed as expected
obj_sds/TEST2: START: obj_sds
 C:\projects\pmdk\src\x64\Debug\tests\obj_sds.exe 1 1 C:\temp\\test_obj_sds2ðŸ˜˜â â â ™â …É—PMDKÓœâ¥ºðŸ™‹\pool.set 5 0 7 0 9 0 11 0 12 0
Mock shutdown_state_is_supported called
Mock shutdown_state_is_supported called
Mock shutdown_state_is_supported called
Mock shutdown_state_is_supported called
Mock shutdown_state_is_supported called
Failed as planned
The second run passed wrongly
obj_sds/TEST2: START: obj_sds
 C:\projects\pmdk\src\x64\Debug\tests\obj_sds.exe 0 0 C:\temp\\test_obj_sds2ðŸ˜˜â â â ™â …É—PMDKÓœâ¥ºðŸ™‹\pool.set 5 1 7 1 9 1 11 0 12 0
obj_sds/TEST2: DONE
NotSpecified: (:) [Write-Error], WriteErrorException
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,fail
    + PSComputerName        : localhost
 
obj_sds/TEST2: FAILEDobj_sds/TEST2: FAILED
    + CategoryInfo          : OperationStopped: (obj_sds/TEST2: FAILED:String) [], RuntimeException
    + FullyQualifiedErrorId : obj_sds/TEST2: FAILED
    + PSComputerName        : localhost
 
C:\projects\pmdk\src\test\RUNTESTS.PS1 : RUNTESTS FAILED: one of the tests failed
At line:14 char:9
+         ./RUNTESTS.ps1 -b debug -o 4m
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,RUNTESTS.PS1
 
Command executed with exception: RUNTESTS FAILED: one of the tests failed
```

pbalcer commented:
I downloaded your code on my PC and reproduced the problem, but I didn't have the chance to debug it yet. I'll do that tomorrow (unless @lplewa wakes up and tells us the answer - it's probably something trivial that we are missing :P)

lucmaga commented:
@pbalcer I finally figure it out after I read the comment on `src/test/unittest/unittest.h` explaining how the windows mocks are working and play a little bit with it.
The problem is that the `.vcxproj` file is forcing to include `mock_windows.h` it is doing so in all source files listed on it for that build. This include will use the preprocessor to replace the name of the functions for the `__wrap_` ones. When I added the
macro to replace `shutdown_state_is_supported` it was doing it on `commom/shutdown_state.c` as well then throwing a
double definition error during compilation. When I removed `commom/shutdown_state.c` from the `.vcxproj` it builds but
was not actually mocking the other methods.
This is hack for mocking functions on windows took some time for me to understand...
Thanks for all the help.

pbalcer commented on src/common/shutdown_state.c:
```diff
@@ -20,6 +20,29 @@
+/*
+ * shutdown_state_is_supported -- (internal) check if device supports SDS
+ *
+ * Returns 0 if supports and 1 if not.
```
I'd probably reverse this.
We usually return 0 for functions that do some processing, but functions that check a condition should just return 0 for false and 1 for true.

lucmaga replied:
Done

pbalcer commented on src/test/libpmempool_feature/TEST2w.PS1:
```diff
@@ -19,6 +19,14 @@ expect_normal_exit $PMEMPOOL create obj $POOL
 # PMEMPOOL_FEAT_SHUTDOWN_STATE is enabled by default
 libpmempool_feature_query "SHUTDOWN_STATE"
 
+# If SDS is not enabled at this pois is because SDS is not available for
```
point

lucmaga replied:
Done

lucmaga commented:
Is there any changes pending for this to be merged?

pbalcer commented:
No, I'm just waiting for @lplewa to review your patch.

lplewa commented:
As i wroten before:

"you should remove this line: (and similar from this file)
https://github.com/pmem/pmdk/blob/master/src/common/shutdown_state.c#L74"

this code should be obsolite after your change.

lucmaga commented:
@pbalcer Can you restart the CI job?
obj_defrag_advanced/TEST3 timed out but I don't think it has anything to do with my changes. Are you seeing this happen?

pbalcer commented:
Restarted.
This sometimes happens unfortunately. We will likely need to change the timeout for this test.

lucmaga commented:
Thanks @pbalcer. It seems I finally got this right but it took many more changes.

I notice that TravisCI is disabled. Is that intentional?

pbalcer commented:
No, travis should be running. Odd...

pbalcer commented:
It looks to be running just fine if you go to directly to travis: https://travis-ci.org/github/pmem/pmdk/builds
It's just not showing up properly in the checks for a PR.

lucmaga commented:
@pbalcer, @lplewa PING!
